Add required and validation checks for input parameters

kvigen opened the pull request:
This change updates the way we handle input parameters in a few ways:
1. It makes non-required fields pointers
2. It errors if a field is required and isn't specified
3. It validates the input parameters based on the swagger spec. Since we don't support array or schema type input parameters, this isn't too difficult

We considered using go-swagger's internal codegen for this, but instead decided it was easier to just use their validation libraries and generate the code ourselves.

rgarcia commented on README.md:
```diff
@@ -140,11 +140,6 @@ make test
-  - Support for the required field in the Swagger spec
-  - At least some of the validation defined in http://swagger.io/specification/#parameterIn. In particular we want to be able to support checking whether an input parameter is a valid object ID.
-- More robust server defaults
-  - SIGTERM handling
```
seeing this diff makes me think that it would be useful to have a list of features at the top of this readme

kvigen replied:
agreed, will add this in a future PR

rgarcia commented on client/genclients.go:
```diff
@@ -176,16 +176,19 @@ func successResponse(outputName string) string {
+	if !p.Required {
+		valToSet = "*" + valToSet
+	}
 	switch p.Type {
```
might be useful in these switch statements to link to the swagger spec that enumerates the different values

rgarcia commented on generated/models/inputs.go:
```diff
@@ -2,34 +2,77 @@ package models
+
+	if err := validate.MultipleOf("maxPages", "query", *i.MaxPages, 0.500000); err != nil {
+		return err
+	}
```
nice

rgarcia commented on generated/server/handlers.go:
```diff
@@ -57,19 +58,24 @@ func GetBooksHandler(ctx context.Context, w http.ResponseWriter, r *http.Request
-	if err != nil {
-		return nil, err
-	}
+	availableTmp, err := strconv.ParseBool(availableStr)
```
looks like this err is never checked

kvigen replied:
when i originally wrote it i thought it made sense to ignore errors if the parameter wasn't required, but now that I think about it, i should error here if they did specify something and it failed

rgarcia commented on generated/server/handlers.go:
```diff
@@ -158,9 +172,10 @@ func CreateBookHandler(ctx context.Context, w http.ResponseWriter, r *http.Reque
+	var err error
+	_ = err
+
+	err = json.NewDecoder(r.Body).Decode(input.NewBook)
```
nit: this code now looks much more auto-generated (I like the previous `if err := json.NewDecoder(...` better)

kvigen replied:
changed it back

rgarcia commented on generated/server/handlers.go:
```diff
@@ -57,19 +58,24 @@ func GetBooksHandler(ctx context.Context, w http.ResponseWriter, r *http.Request
-	if err != nil {
-		return nil, err
-	}
+	maxPagesTmp, err := strconv.ParseFloat(maxPagesStr, 64)
```
same here

rgarcia commented:
I think I've mentioned this but one other swagger feature that I sort of lump together with input validation is their "data type formats": http://swagger.io/specification/#dataTypeFormat Most useful of which is `date-time`. It'd be great to add support for this in this PR, but if it's too much can you add it as a future feature?

kvigen commented:
derp, i meant to add the "data type formats" here and just forgot to do it. I think it should be pretty straightforward. Just a new parse / validate function in the switch statements:
`formats.Parse("date", raw)`
`validate.FormatOf("district_id", "query", "date", string(o.DistrictID), formats)`

I will update with all 5 that are missing.

kvigen commented:
@rgarcia I added the formats, except I left off binary for now since it converted to an `io.Reader` / `io.Writer` and I didn't really want to deal with it

I also refactored the code quite a bit, so this because a bit bigger change than I originally anticipated. I put all the parameter validation logic into a separate file, which I think makes it a bit clearer what we support, and how there are 4 common operations on the types. I also added quite a few `if x != nil` checks where I probably should have added them before. If this PR has too much going on, let me know and I can try to break it up.

There are still a few things I could do to clean up, but I figured this was enough for one PR.

rgarcia commented on generated/models/inputs.go:
```diff
@@ -2,34 +2,116 @@ package models
 
 func (i GetBooksInput) Validate() error {
+	if i.State != nil {
+
```
nit: blank line

rgarcia commented on swagger/parameter.go:
```diff
@@ -0,0 +1,238 @@
+		pointer = "*"
+	}
+	return fmt.Sprintf("%si.%s", pointer, Capitalize(param.Name))
+}
```
:clap: I like splitting this out

rgarcia commented:
lgtm!
